docs/design/2020-06-24-placement-rules-in-sql.md

```diff
@@ -327,255 +388,102 @@ In Geo-Partitioning, the table must be splitted into partitions, and each partit
+### Storing Placement Policies
 
-## Implementation
+Placement policies will be stored in a table in the `mysql` schema. The policy name must be globally unique, but the definition of the table is not described in this proposal.
 
-This section focuses on the implemention details of defining placement rules in SQL.
+Because the implementation is TiDB specific (does not require any compatibility with MySQL), it is up to the implementer to decide.
 
 ### Storing placement rules
 
-PD uses placement rules to schedule data, so a replica of placement rules must be persistent on the PD. 
+PD uses placement rules to schedule data, so a replica of placement rules for _tables and partitions_ must be persistent in PD. 
 
 However, TiDB also uses placement rules in some cases, as discussed in section "Querying placement rules". There are basically 2 ways to achieve this:
 
 - Save the placement rules in table information, which will be duplicated with PD
 - Only PD persists the placement rules, while TiDB caches a copy of them
 
-Before choosing the solution, transactional requirements need to be noticed:
-
-- Defining placement rules may fail, and users will probably retry it. As retrying `ADD PLACEMENT POLICY` will add more replicas than expected, the atomicity of the opertion needs to be guaranteed.
-- `ADD PLACEMENT POLICY` needs to read the original placement rules, combine the 2 rules and then store them to PD, so linearizability should be gauranteed.
+Defining placement rules may fail, but the handling of this is simplified from earlier proposals. Because placement rules are specified in full (i.e. there is no longer syntax for `ALTER TABLE .. ADD PLACEMENT POLICY`), it is safe to automatically retry applying rules.
 
 If the placement rules are stored on both TiKV and PD, the approaches to keep atomicity are as follows:
 
 - Enforce a 2PC protocol on TiKV and PD.
 - Store them on TiKV along with a middle state. If TiKV succeeds, then try PD, otherwise rollback it by the middle state. The DDL procedure guarantees the atomicity even if TiDB is down.
 
-The approaches to keep linearizability are as follows:
-
-- Define placement rules in serial.
-- Enforce an exclusive lock on one of the replicas and release it after the whole job finishes.
-
 As a contrast, if the placement rules are stored only on PD, the approaches to keep atomicity are as follows:
 
 - Write all the placement rules in one ETCD transaction.
 - Persist a middle state on TiKV before sending to PD. This middle state acts as undo log.
 
-The approaches to keep linearizability are as follows:
-
-- Define placement rules in serial.
-- Enforce an exclusive lock on PD and release it after the job finishes.
-
 The comparison shows that both solutions are possible, but storing placement rules only on PD is more practical. To guarantee the transactional characteristics, the easiest way is to write all placement rules in a transaction and define them in serial on the TiDB side.
```

Comment:
But it seems storing placement rules as part of infoschema data in tikv is more easy to make strong consistency with other metas like placement policy. After that we don't have to sync from PD for cases like "show create table". I think it is not hard to implement 2PC using current ddl change algorithm.

Reply:
At the early stage of this project, it is hard, for PD API won't rollback, i.e. non-atomic. You will need to use TiDB to emulate transaction-like behaviors, which is not safe.

But, now, PD itself will rollback automatically in one HTTP request. So, yes, the thing changes: we can accept DDL middle states and DDL persistence on TiKV.

Reply:
What do you mean by "rollback automatically in one HTTP request" ?

Reply:
Say I post 4 rules to PD, if 3rd rule failed to apply, 1st/2nd rules will be applied anyway. So TiDB must rollback them.
2pc HTTP API for PD is not elegant, too, and they prefer rest API.
And one more thing, we can not post multiple rules in one request at the time, which makes the management of state even harder.

As a result, persisted on TiKV, and trying to manage state of PD was a disaster at the time. Now, the introduction of batch api/ rule bundles changed the game.

Reply:
I have clarified this better under implementation. PTAL again and let me know if it needs more details.